tf2_ros/include/tf2_ros/message_filter.h

```diff
@@ -494,27 +495,37 @@ class MessageFilter : public MessageFilterBase, public message_filters::SimpleFi
     // find the message this request is associated with
     typename L_MessageInfo::iterator msg_it = messages_.begin();
     typename L_MessageInfo::iterator msg_end = messages_.end();
-    for (; msg_it != msg_end; ++msg_it) {
-      MessageInfo & info = *msg_it;
-      auto handle_it = std::find(info.handles.begin(), info.handles.end(), handle);
-      if (handle_it != info.handles.end()) {
-        // found msg_it
-        ++info.success_count;
-        break;
+
+    MessageInfo saved_msg;
+
+    {
+      // We will be accessing and mutating messages now, require unique lock
+      std::unique_lock<std::mutex> lock(messages_mutex_);
+
+      for (; msg_it != msg_end; ++msg_it) {
+        MessageInfo & info = *msg_it;
+        auto handle_it = std::find(info.handles.begin(), info.handles.end(), handle);
+        if (handle_it != info.handles.end()) {
+          // found msg_it
+          ++info.success_count;
+          if (info.success_count >= expected_success_count_) {
+            saved_msg = *msg_it;
+            messages_.erase(msg_it);
+            --message_count_;
+          } else {
+            return;
```

Comment:
From a purely stylistic perspective, I find this hard to follow.  Instead, I'll suggest we do:

```
msg_it = msg_end;
```

And then let it fall through to hit the break, which will then cause it to return in the conditional below.

Alternatively, we could move the break inside of the `if` brackets, which would also be slightly easier to reason about in my opinion.